[C10D] Implement Store fallbacks for append, multi_get and multi_set.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100768
* #100742

These fallbacks exposed some issue in quite a few spots in our bindings.

kumpera committed:
[C10D] Implement Store fallbacks for append, multi_get and multi_set.

These fallbacks exposed some issue in quite a few spots in our bindings.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement Store fallbacks for append, multi_get and multi_set."


These fallbacks exposed some issue in quite a few spots in our bindings.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/Store.cpp:
```diff
@@ -38,18 +38,34 @@ std::string Store::get_to_str(const std::string& key) {
+  std::vector<uint8_t> current;
+  // cannot use get(key) as it might block forever if the key doesn't exist
+  current = compareSet(key, current, expected);
+  while (current != expected) {
+    expected = current;
+    expected.insert(expected.end(), value.begin(), value.end());
+    current = compareSet(key, current, expected);
+  }
```
How can we prevent forever loop?

kumpera replied:
In theory, we can't prevent infinite loops when using CAS as long as there are multiple active clients going at each other.

OTOH, we don't have spurious, no progress, CAS failures so each time it fails there has being global progress.

fduwjj commented on torch/csrc/distributed/c10d/Store.cpp:
```diff
@@ -38,18 +38,34 @@ std::string Store::get_to_str(const std::string& key) {
 }
 
 void Store::append(const std::string& key, const std::vector<uint8_t>& value) {
```
This is a fallback when `has_extended_api` is false right?

kumpera replied:
Yup, these fallbacks are in place when the underlying store doesn't implement the newer methods.

fduwjj reviewed (approved):
Unblock.

kumpera committed:
Update on "[C10D] Implement Store fallbacks for append, multi_get and multi_set."


These fallbacks exposed some issue in quite a few spots in our bindings.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement Store fallbacks for append, multi_get and multi_set."


These fallbacks exposed some issue in quite a few spots in our bindings.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge
